distributed/spill.py

```diff
@@ -49,28 +82,67 @@ def disk(self) -> Mapping[Hashable, Any]:
         """
         return self.slow
 
-    @staticmethod
-    def _weight(key: Hashable, value: Any) -> int:
-        return safe_sizeof(value)
-
-    def _on_evict(self, key: Hashable, value: Any) -> None:
-        b = safe_sizeof(value)
-        self.spilled_by_key[key] = b
-        self.spilled_total += b
-
-    def _on_retrieve(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key)
-
-    def __setitem__(self, key: Hashable, value: Any) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
-        super().__setitem__(key, value)
-        if key in self.slow:
-            # value is individually larger than target so it went directly to slow.
-            # _on_evict was not called.
-            b = safe_sizeof(value)
-            self.spilled_by_key[key] = b
-            self.spilled_total += b
-
-    def __delitem__(self, key: Hashable) -> None:
-        self.spilled_total -= self.spilled_by_key.pop(key, 0)
+    @property
+    def spilled_total(self) -> int:
+        return self.slow.total_weight
+
+
+def _in_memory_weight(key: Hashable, value: Any) -> int:
+    return safe_sizeof(value)
+
+
+class MaxSpillExceeded(Exception):
+    pass
+
+
+class PickleError(Exception):
+    pass
+
+
+class Slow(zict.Func):
+    max_weight: int | Literal[False]
+    weight_by_key: dict[Hashable, int]
+    total_weight: int
+
+    def __init__(self, spill_directory: str, max_weight: int | Literal[False] = False):
+        super().__init__(
+            partial(serialize_bytelist, on_error="raise"),
+            deserialize_bytes,
+            zict.File(spill_directory),
+        )
+        self.max_weight = max_weight
+        self.weight_by_key = {}
+        self.total_weight = 0
+
+    def __setitem__(self, key, value):
+        try:
+            pickled = self.dump(value)
+            pickled_size = sum(len(frame) for frame in pickled)
+        except Exception as e:  # This is the error raised when fail to serialize
+            raise PickleError(
+                key, e
+            )  # It's ok to loose the bad key, cluster remains usable
```

Comment:
```suggestion
        except Exception as e:
            # zict.LRU ensures that the key remains in fast if we raise.
            # Wrap the exception so that it's recognizable by SpillBuffer,
            # which will then unwrap it.
            raise PickleError(key, e)
            
        pickled_size = sum(len(frame) for frame in pickled)
```